gpflow/conditionals/util.py

```diff
@@ -51,6 +52,18 @@ def base_conditional(Kmn: tf.Tensor,
         0)  # [N]
     Kmn = tf.transpose(Kmn, perm)  # [..., M, N]
 
+    shape_constraints = [
+        (Kmn, [..., 'M', 'N']),
+        (Kmm, ['M', 'M']),
+        (Knn, [..., 'N', 'N'] if full_cov else [..., 'N']),
+        (f, ['M', 'R']),
+    ]
+    if q_sqrt is not None:
+        shape_constraints.append(
+            (q_sqrt, (['M', 'R'] if q_sqrt.shape.ndims == 2 else ['R', 'M', 'M']))
+        )
+    assert_shapes(shape_constraints)
```

Comment:
`tf.debugging.assert_shapes` is still good for function arguments, while `utilities.assert_shapes` is for internal code consistency checks

Reply:
I think we should consistently use one of them, not mix&match tf.debugging.assert_shapes and our wrapper utilities.assert_shapes. If we are saying we want our own wrapper to not always raise ValueErrors, wouldn't it be better to simply give it an argument that says whether to replace ValueError with ShapeError or not?

Reply:
yeah could do that

Reply:
what's wrong with always raising a ShapeError though ?

Reply:
It's a good question, and I'm not entirely clear on the reasons, but essentially I believe it boils down to the fact that the two errors have occurred for different reasons. See [this question](https://stackoverflow.com/q/117171/5986907) for popular answers.

The error at the beginning of the function is a result of erroneous user input, which could have resulted from any number of things out of our control. Perhaps a malformed CSV has been uploaded to some ML pipeline and it's missing a column. Crucially, it could be that they can recover from that situation.

The error at the end of the function is a check of the GPflow code. If there's a bug in GPflow, the user can't do anything about that.